File: cpp/src/v0/pipeline.cpp
#include "pipeline.hpp"

#include "matching_engine.hpp"

#include <algorithm>
#include <chrono>
#include <sstream>

namespace ytta::v0 {
namespace {

int type_rank(GoldenType t) {
  switch (t) {
    case GoldenType::Md:
      return 0;
    case GoldenType::Action:
      return 1;
    case GoldenType::Ack:
      return 2;
    case GoldenType::Fill:
      return 3;
  }
  return 99;
}

}  // namespace

std::string format_action(const StrategyAction& a) {
  std::ostringstream os;
  os << "{\"type\":\"action\",\"ts_ns\":" << a.ts_ns << ",\"action\":\"";
  switch (a.kind) {
    case ActionKind::Noop:
      os << "NOOP";
      break;
    case ActionKind::NewOrder:
      os << "NEW_ORDER";
      break;
    case ActionKind::Cancel:
      os << "CANCEL";
      break;
  }
  os << "\"";
  if (a.kind == ActionKind::NewOrder) {
    os << ",\"cl_ord_id\":" << a.cl_ord_id << ",\"side\":\"" << to_string(a.side)
       << "\",\"price\":" << a.price << ",\"qty\":" << a.qty;
  } else if (a.kind == ActionKind::Cancel) {
    os << ",\"cl_ord_id\":" << a.cl_ord_id;
  }
  os << ",\"reason\":\"" << a.reason << "\"}";
  return os.str();
}

std::string format_engine_event(const EngineEvent& e) {
  std::ostringstream os;
  if (const auto* ack = std::get_if<AckEvent>(&e)) {
    os << "{\"type\":\"ack\",\"ts_ns\":" << ack->ts_ns
       << ",\"cl_ord_id\":" << ack->cl_ord_id << ",\"status\":\""
       << to_string(ack->status) << "\"";
    if (ack->status == AckStatus::Rejected) {
      os << ",\"reason\":\"" << ack->reason << "\"";
    }
    os << "}";
    return os.str();
  }
  if (const auto* fill = std::get_if<FillEvent>(&e)) {
    os << "{\"type\":\"fill\",\"ts_ns\":" << fill->ts_ns
       << ",\"cl_ord_id\":" << fill->cl_ord_id << ",\"price\":" << fill->price
       << ",\"qty\":" << fill->qty << ",\"liquidity\":\""
       << to_string(fill->liquidity) << "\"}";
    return os.str();
  }
  const auto& md = std::get<MdEvent>(e);
  os << "{\"type\":\"md\",\"ts_ns\":" << md.ts_ns << ",\"instrument\":\""
     << md.instrument << "\",\"best_bid\":" << md.best_bid
     << ",\"best_ask\":" << md.best_ask << ",\"best_bid_qty\":" << md.best_bid_qty
     << ",\"best_ask_qty\":" << md.best_ask_qty << "}";
  return os.str();
}

void sort_golden(std::vector<GoldenEvent>& events) {
  std::sort(events.begin(), events.end(),
            [](const GoldenEvent& a, const GoldenEvent& b) {
              if (a.ts_ns != b.ts_ns) {
                return a.ts_ns < b.ts_ns;
              }
              const int ra = type_rank(a.type);
              const int rb = type_rank(b.type);
              if (ra != rb) {
                return ra < rb;
              }
              return a.cl_ord_id < b.cl_ord_id;
            });
}

std::vector<GoldenEvent> Pipeline::run(const std::vector<Tick>& ticks,
                                       LatencyProbe& probe) {
  MatchingEngine engine;
  OrderGateway gateway(engine);
  Strategy strategy;
  std::vector<GoldenEvent> out;

  for (const auto& tick : ticks) {
    const auto t0 = std::chrono::steady_clock::now();
    const StrategyAction action = strategy.on_tick(tick);

    GoldenEvent ge;
    ge.ts_ns = action.ts_ns;
    ge.type = GoldenType::Action;
    ge.cl_ord_id =
        (action.kind == ActionKind::Noop) ? 0 : action.cl_ord_id;
    ge.line = format_action(action);
    out.push_back(std::move(ge));

    if (action.kind != ActionKind::Noop) {
      std::vector<EngineEvent> events;
      if (action.kind == ActionKind::NewOrder) {
        events = gateway.new_order(
            Order{action.cl_ord_id, action.side, action.price, action.qty},
            action.ts_ns);
      } else {
        events = gateway.cancel(action.cl_ord_id, action.ts_ns);
      }
      for (const auto& ev : events) {
        GoldenEvent ee;
        ee.ts_ns = action.ts_ns;
        ee.line = format_engine_event(ev);
        if (const auto* ack = std::get_if<AckEvent>(&ev)) {
          ee.type = GoldenType::Ack;
          ee.cl_ord_id = ack->cl_ord_id;
        } else if (const auto* fill = std::get_if<FillEvent>(&ev)) {
          ee.type = GoldenType::Fill;
          ee.cl_ord_id = fill->cl_ord_id;
        } else {
          ee.type = GoldenType::Md;
          ee.cl_ord_id = 0;
        }
        out.push_back(std::move(ee));
      }
    }

    probe.record_ns(steady_ns_since(t0));
  }

  sort_golden(out);
  return out;
}

}  // namespace ytta::v0

File: rust/src/v0/pipeline.rs
use std::time::Instant;

use super::engine::MatchingEngine;
use super::gateway::OrderGateway;
use super::probe::LatencyProbe;
use super::strategy::Strategy;
use super::types::{
    ActionKind, EngineEvent, Order, StrategyAction, Tick,
};

#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
pub enum GoldenType {
    Md = 0,
    Action = 1,
    Ack = 2,
    Fill = 3,
}

#[derive(Debug, Clone)]
pub struct GoldenEvent {
    pub ts_ns: i64,
    pub type_: GoldenType,
    pub cl_ord_id: u64,
    pub line: String,
}

pub fn format_action(a: &StrategyAction) -> String {
    let mut s = format!(
        "{{\"type\":\"action\",\"ts_ns\":{},\"action\":\"",
        a.ts_ns
    );
    match a.kind {
        ActionKind::Noop => s.push_str("NOOP"),
        ActionKind::NewOrder => s.push_str("NEW_ORDER"),
        ActionKind::Cancel => s.push_str("CANCEL"),
    }
    s.push('"');
    match a.kind {
        ActionKind::NewOrder => {
            s.push_str(&format!(
                ",\"cl_ord_id\":{},\"side\":\"{}\",\"price\":{},\"qty\":{}",
                a.cl_ord_id,
                a.side.as_str(),
                a.price,
                a.qty
            ));
        }
        ActionKind::Cancel => {
            s.push_str(&format!(",\"cl_ord_id\":{}", a.cl_ord_id));
        }
        ActionKind::Noop => {}
    }
    s.push_str(&format!(",\"reason\":\"{}\"}}", a.reason));
    s
}

pub fn format_engine_event(e: &EngineEvent) -> (GoldenType, u64, String) {
    match e {
        EngineEvent::Ack(ack) => {
            let mut line = format!(
                "{{\"type\":\"ack\",\"ts_ns\":{},\"cl_ord_id\":{},\"status\":\"{}\"",
                ack.ts_ns,
                ack.cl_ord_id,
                ack.status.as_str()
            );
            if ack.status == super::types::AckStatus::Rejected {
                line.push_str(&format!(",\"reason\":\"{}\"", ack.reason));
            }
            line.push('}');
            (GoldenType::Ack, ack.cl_ord_id, line)
        }
        EngineEvent::Fill(fill) => {
            let line = format!(
                "{{\"type\":\"fill\",\"ts_ns\":{},\"cl_ord_id\":{},\"price\":{},\"qty\":{},\"liquidity\":\"{}\"}}",
                fill.ts_ns,
                fill.cl_ord_id,
                fill.price,
                fill.qty,
                fill.liquidity.as_str()
            );
            (GoldenType::Fill, fill.cl_ord_id, line)
        }
        EngineEvent::Md(md) => {
            let line = format!(
                "{{\"type\":\"md\",\"ts_ns\":{},\"instrument\":\"{}\",\"best_bid\":{},\"best_ask\":{},\"best_bid_qty\":{},\"best_ask_qty\":{}}}",
                md.ts_ns, md.instrument, md.best_bid, md.best_ask, md.best_bid_qty, md.best_ask_qty
            );
            (GoldenType::Md, 0, line)
        }
    }
}

pub fn sort_golden(events: &mut [GoldenEvent]) {
    events.sort_by(|a, b| {
        a.ts_ns
            .cmp(&b.ts_ns)
            .then(a.type_.cmp(&b.type_))
            .then(a.cl_ord_id.cmp(&b.cl_ord_id))
    });
}

pub fn run_pipeline(ticks: &[Tick], probe: &mut LatencyProbe) -> Vec<GoldenEvent> {
    let mut engine = MatchingEngine::new();
    let mut strategy = Strategy::new();
    let mut out = Vec::new();

    for tick in ticks {
        let t0 = Instant::now();
        let action = strategy.on_tick(tick);

        let cl = if action.kind == ActionKind::Noop {
            0
        } else {
            action.cl_ord_id
        };
        out.push(GoldenEvent {
            ts_ns: action.ts_ns,
            type_: GoldenType::Action,
            cl_ord_id: cl,
            line: format_action(&action),
        });

        if action.kind != ActionKind::Noop {
            let mut gw = OrderGateway::new(&mut engine);
            let events = match action.kind {
                ActionKind::NewOrder => gw.new_order(
                    Order {
                        cl_ord_id: action.cl_ord_id,
                        side: action.side,
                        price: action.price,
                        qty: action.qty,
                    },
                    action.ts_ns,
                ),
                ActionKind::Cancel => gw.cancel(action.cl_ord_id, action.ts_ns),
                ActionKind::Noop => unreachable!(),
            };
            for ev in events {
                let (gt, id, line) = format_engine_event(&ev);
                out.push(GoldenEvent {
                    ts_ns: action.ts_ns,
                    type_: gt,
                    cl_ord_id: id,
                    line,
                });
            }
        }

        probe.record_ns(t0.elapsed().as_nanos() as i64);
    }

    sort_golden(&mut out);
    out
}
